src/hivpy/circumcision.py

```diff
@@ -0,0 +1,182 @@
+import importlib.resources
+from datetime import timedelta
+
+import numpy as np
+
+import hivpy.column_names as col
+
+from .circumcision_data import CircumcisionData
+from .common import SexType, rng
+
+
+class CircumcisionModule:
+
+    def __init__(self, **kwargs):
+        # init cirumcision data
+        with importlib.resources.path("hivpy.data", "circumcision.yaml") as data_path:
+            self.c_data = CircumcisionData(data_path)
+        self.mc_int = self.c_data.mc_int
+        self.year_interv = self.c_data.year_interv
+        self.test_link_circ_prob = self.c_data.test_link_circ_prob
+        self.circ_inc_rate_year_i = self.c_data.circ_inc_rate_year_i
+        # NOTE: the covid disrup field may not belong here
+        self.covid_disrup_affected = self.c_data.covid_disrup_affected
+        self.vmmc_disrup_covid = self.c_data.vmmc_disrup_covid
+        self.circ_inc_rate = self.c_data.circ_inc_rate.sample()
+        self.rel_incr_circ_post_2013 = self.c_data.rel_incr_circ_post_2013.sample()
+        self.circ_inc_15_19 = self.c_data.circ_inc_15_19.sample()
+        self.circ_red_20_30 = self.c_data.circ_red_20_30.sample()
+        self.circ_red_30_50 = self.c_data.circ_red_30_50.sample()
+        self.prob_birth_circ = self.c_data.prob_birth_circ.sample()
+
+    def init_birth_circumcision_all(self, population, date):
+        """
+        Initialise circumcision at birth for the entire male population,
+        both born and unborn. COVID disruption is not factored in.
+        """
+        male_population = population.index[population[col.SEX] == SexType.Male]
+        r = rng.uniform(size=len(male_population))
+        circumcision = r < self.prob_birth_circ
+        population.loc[male_population, col.CIRCUMCISED] = circumcision
+        # split newly circumcised population into born and unborn
+        circ_born_population = population.index[population[col.CIRCUMCISED]
+                                                & (population[col.AGE] >= 0.25)]
+        circ_unborn_population = population.index[population[col.CIRCUMCISED]
+                                                  & (population[col.AGE] < 0.25)]
+        # use current simulation start date as circumcision date for born individuals
+        population.loc[circ_born_population, col.CIRCUMCISION_DATE] = date
+        # find date where each unborn individual's age would be 0.25
+        population.loc[circ_unborn_population,
+                       col.CIRCUMCISION_DATE] = population[col.AGE].transform(
+                                                lambda x: date - timedelta(days=(x+0.25)*365))
```

Comment:
I think this should be `(x-0.25)` because `x` is their (currently negative) age, so the time until they get circumcised is $(-x) + 0.25$ or $- (x-0.25)$